code/modules/mob/dead/observer/orbit.dm

```diff
@@ -18,94 +34,132 @@
-	for (var/name in pois)
+	for(var/name in new_pois)
 		var/list/serialized = list()
 		serialized["name"] = name
 
-		var/poi = pois[name]
+		var/poi = new_pois[name]
 
-		serialized["ref"] = REF(poi)
+		var/poi_ref = REF(poi)
+		serialized["ref"] = poi_ref
 
-		var/mob/M = poi
-		if (istype(M))
-			if (isobserver(M))
-				var/number_of_orbiters = length(M.get_all_orbiters())
+		pois[poi_ref] = poi
+
+		if(ismob(poi))
+			var/mob/mob_poi = poi
+
+			if(isobserver(mob_poi))
+				var/number_of_orbiters = length(mob_poi.get_all_orbiters())
 				if (number_of_orbiters)
 					serialized["orbiters"] = number_of_orbiters
 				ghosts += list(serialized)
-			else if (M.stat == DEAD)
+				continue
+
+			if(mob_poi.stat == DEAD)
 				dead += list(serialized)
-			else if (M.mind == null)
+				continue
+
+			if(isnull(mob_poi.mind))
 				npcs += list(serialized)
-			else
-				var/number_of_orbiters = length(M.get_all_orbiters())
-				if (number_of_orbiters)
-					serialized["orbiters"] = number_of_orbiters
+				continue
 
-				var/datum/mind/mind = M.mind
-				var/was_antagonist = FALSE
+			var/number_of_orbiters = length(mob_poi.get_all_orbiters())
+			if(number_of_orbiters)
+				serialized["orbiters"] = number_of_orbiters
 
-				for (var/_A in mind.antag_datums)
-					var/datum/antagonist/A = _A
-					if (A.show_to_ghosts)
-						was_antagonist = TRUE
-						serialized["antag"] = A.name
-						antagonists += list(serialized)
-						break
+			var/datum/mind/mind = mob_poi.mind
+			var/was_antagonist = FALSE
 
-				if (!was_antagonist)
-					alive += list(serialized)
+			for(var/datum/antagonist/antag_datum as anything in mind.antag_datums)
+				if (antag_datum.show_to_ghosts)
+					was_antagonist = TRUE
+					serialized["antag"] = antag_datum.name
+					antagonists += list(serialized)
+					break
+
+			if(!was_antagonist)
+				alive += list(serialized)
 		else
 			misc += list(serialized)
 
-	data["alive"] = alive
-	data["antagonists"] = antagonists
-	data["dead"] = dead
-	data["ghosts"] = ghosts
-	data["misc"] = misc
-	data["npcs"] = npcs
-	return data
-
-/datum/orbit_menu/ui_assets()
-	. = ..() || list()
-	. += get_asset_datum(/datum/asset/simple/orbit)
+	do_update_next_process = TRUE
+
+/// Shows the UI to the specified user.
+/datum/orbit_menu/proc/show(mob/user)
+	ui_interact(user)
+
+/**
+ * Signal handler for major POI changes.
+ *
+ * Major POI changes are POIs added or removed from SSpoi's lists.
+ *
+ * We need to update the poi lists promptly, as they are used to validate user input in ui_act.
+ */
+/datum/orbit_menu/proc/on_poi_change()
+	SIGNAL_HANDLER
+
+	update_poi_list()
+
+/**
+ * Manages periodic updates.
+ *
+ * Used to sweep up minor POI updates like dead < - > alive /mob/living state changes.
+ *
+ * Pushes all the updates to all active UI windows.
+ */
+/datum/orbit_menu/process(delta_time)
+	update_poi_list()
+
+	for(var/datum/tgui/window as anything in SStgui.open_uis_by_src[REF(src)])
+		window.send_full_update()
```

Comment:
If you're doing this, change it to `ui_data()`, but I specifically avoided this auto update for a reason, it's fairly slow.

I would very much recommend not doing this.

Reply:
Moved to ui_data. In order to not take the massive hit constantly sorting the POI lists for every player with the interface open, the lists in SSpoints_of_interest are pre-sorted and use a slightly modified BINARY_INSERT that lets me slip in a proc name instead of a var name to compare. New binary insert is tested and working (see Testing section in PR body).